x/oracle/keeper/task.go

```diff
@@ -110,72 +107,109 @@ func (k Keeper) IteratorClosingTaskIDsByTime(ctx sdk.Context, endTime time.Time,
+				tk := protoTids[i].Tid
+				if store.Has(tk) {
+					store.Delete(tk)
+				}
+			}
+			store.Delete(key)
+			return false
+		})
 }
 
-// CreateTask creates a new task.
+func (k Keeper) GetInvalidTaskIDs(ctx sdk.Context) (resIDs []types.TaskID) {
+	resIDs = append(resIDs, k.GetClosingTaskIDsByHeight(ctx, ctx.BlockHeight())...)
+	k.IteratorTaskIDsByEndTime(
+		ctx, types.ClosingTaskStoreKeyTimedPrefix, ctx.BlockTime(),
+		func(key, value []byte) bool {
+			var taskIDsProto types.TaskIDs
+			k.cdc.MustUnmarshalLengthPrefixed(value, &taskIDsProto)
+			resIDs = append(resIDs, taskIDsProto.TaskIds...)
+			return false
+		})
+	return resIDs
+}
+
+// calling of CreateTask creates one of following
+// 1. Task (smart contract task)
+// 2. TxTask (transaction task)
+// 3. placeholder TxTask (status:TaskStatusNil, creator: nil, bounty:nil)
 func (k Keeper) CreateTask(ctx sdk.Context, creator sdk.AccAddress, task types.TaskI) error {
 	savedTask, err := k.GetTask(ctx, task.GetID())
 	if err == nil {
-		if _, ok := task.(*types.Task); ok {
-			if savedTask.IsValid(ctx) {
-				return types.ErrTaskNotClosed
-			}
-			if err := k.DeleteTask(ctx, savedTask); err != nil {
-				return err
-			}
-		} else if reqTask, ok := task.(*types.TxTask); ok {
-			oldTask, ok := savedTask.(*types.TxTask)
-			if !ok {
-				return types.ErrInvalidTask
-			}
-			if !reqTask.IsValid(ctx) {
-				return types.ErrInvalidTask
-			}
-
-			if oldTask.GetStatus() != types.TaskStatusNil {
-				if savedTask.IsValid(ctx) {
-					return types.ErrTaskNotClosed
-				}
-			} else {
-				//created by fast path
-				reqTask.Responses = oldTask.Responses
-				reqTask.Expiration = oldTask.Expiration
-				reqTask.Score = oldTask.Score
-			}
-			if err := k.DeleteTask(ctx, savedTask); err != nil {
-				return err
-			}
+		// be noted that a TaskStatusNil task is not valid
+		if savedTask.IsValid(ctx) {
+			return types.ErrTaskNotClosed
+		}
+		if err := k.DeleteTask(ctx, savedTask); err != nil {
+			return err
 		}
 	}
 
 	k.SetTask(ctx, task)
-	k.SetClosingBlockStore(ctx, task)
-	if err := k.CollectBounty(ctx, task.GetBounty(), creator); err != nil {
-		return err
+	// if task's status is TaskStatusNil, it will not go to ClosingBlockStore,
+	// therefor will not be handled in EndBlocker
+	if task.GetStatus() == types.TaskStatusPending {
+		k.SetClosingBlockStore(ctx, task)
+		if err := k.CollectBounty(ctx, task.GetBounty(), creator); err != nil {
+			return err
+		}
 	}
 	return nil
 }
 
+func (k Keeper) BuildTxTaskWithExpire(ctx sdk.Context, txHash []byte, creator string, bounty sdk.Coins, validTime time.Time, status types.TaskStatus) *types.TxTask {
+	taskParams := k.GetTaskParams(ctx)
+	txTask := types.NewTxTask(txHash, creator, bounty, validTime, status)
+	txTask.Expiration = ctx.BlockTime().Add(taskParams.ExpirationDuration)
+
+	ids := k.GetTaskIDsByTime(ctx, types.ExpireTaskStoreKeyPrefix, txTask.Expiration)
+	ids = append(ids, types.TaskID{Tid: txTask.GetID()})
```

Comment:
you probably don't want to do this since this will mean you load/store excessive ids every time you deal with the end time. Setting the key with the time epoch should work better.

Reply:
Could you shed more light on "Setting the key with the time epoch"?

My personal opinion is that the task ids that have the same expiration time should be grouped together. I'm not quite sure how to avoid one extra reading.

Reply:
@yoongbok-lee do you mean using `time epoch + id` as a key?

Reply:
filed https://github.com/shentufoundation/shentu/issues/630 to track this

Reply:
Thanks @xiumingdou . I will just merge this PR. If @yoongbok-lee has some opinion on this stuff, we could talk and make changes in the future.